[RSDK-8272] use the current power to determine the starting power of goForInternal

JohnN193 opened the pull request:
https://viam.atlassian.net/browse/RSDK-8272

changes to encoded motor and controlled motor to handle repeated SetRPM calls. This behavior occurs often when using motion and would result in the motor running at a significantly lower speed than expected. Additionally stops the loop when stop is called to prevent any "kicks" after a motor is stopped & started.

completed tests:

SetVelocity - sensorcontrolled base - works as expected
SetVelocity - wheeled base - works as expected
SetVelocity repeated commands - sensorcontrolled base - works as expected
SetVelocity repeated commands  - wheeled base - works as expected
encoded GoFor - works as expected
controlled GoFor - works as expected
SetRPM encoded - works as expected(no longer resets back to 20% power with repeated calls)
SetRPM controlled - works as expected(no longer resets back to 20% power with repeated calls)
MoveStraight sensorbase - works as expected
MoveStraight wheeled base - works as expected
MoveStraight sensorbase - works as expected
Spin wheeled base - works as expected

Motion Tests:
sensor controlled base: works
encoded motor wheeled base: works(no longer moves at ~30% motor effort)
controlled motor wheeled base: works

martha-johnston commented on components/motor/gpio/controlled.go:
```diff
@@ -178,7 +178,8 @@ func (cm *controlledMotor) Stop(ctx context.Context, extra map[string]interface{
 	if cm.loop != nil && !cm.loop.GetTuning(ctx) {
-		cm.loop.Pause()
+		cm.loop.Stop()
+		cm.loop = nil
```
I feel like this was kind of the whole point of adding the pause/resume functionality, is there a way to keep this? if not I'm fine with removing it but maybe also look at where else this is being used and remove it if it's no longer useful

JohnN193 replied:
so the thing I really want is to get blocks to reset() when pause is called. gonna try to figure out if thats a thing I can do

JohnN193 replied:
tho we do call reset on the pid blocks, so maybe theres a different block we have to reset? or some other state thats not getting reset

martha-johnston replied:
yeah maybe. If the solution is just to reset all blocks then I think we can go back to creating a new loop every time

JohnN193 replied:
creating a new loop technically works, but i think I can get away with just redoing the constant blocks. Calling reset on those doesn't actually do what we want

randhid replied:
I agree with @martha-johnston here, but maybe we added the wrong helper functions and functionality for our drivers to be spammed with SetPoints. 

So, in vanilla controls, I would expect `UpdateConfig` to give me what I want when I request a new SetPoint to Trapezoidal profile. We have ways of storing the state in the summation block too.

Also - it may be worth it to address the encoded motor's issues first and work out a good solution for the control stuff - it could use a more general look since we've been messing with the code so much.

JohnN193 commented on control/trapezoid_velocity_profile.go:
```diff
@@ -122,6 +122,10 @@ func (s *trapezoidVelocityGenerator) reset() error {
 		s.kppGain = 0.45
 	}
+	s.lastVelCmd = 0
+	if s.cfg.Attribute.Has("init_vel") {
+		s.lastVelCmd = s.cfg.Attribute["init_vel"].(float64)
+	}
```
another option is to make this a boolean to remember the speed.

randhid replied:
This is good enough for now - but I feel like it's a band-aid, we should be able to interrupt trapezoidal profiles with new speed asks, it's a pretty standard feedback control feature to rev up to a new set point.

randhid replied:
Also, we could get a zero force controller out of this package two years ago, so I think something regressed in its functionality, I can't wait for rover QA.

randhid replied:
Hmmm, I think like we should be able to pass the old state while the loop was paused through the `Signal` to `lastvelcommand`. This might not be the right place to deal with this, but I'm okay with a follow up ticket.

JohnN193 replied:
I think I'm just going to tweak the code to explicitly do this, rather than having the roundabout configuration steps

randhid commented:
Adding Nico since I'm OOO next week.

JohnN193 commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -227,6 +227,10 @@ func (sb *sensorBase) Stop(ctx context.Context, extra map[string]interface{}) er
 	sb.opMgr.CancelRunning(ctx)
 	if sb.loop != nil {
 		sb.loop.Pause()
+		// update pid controllers to be an at rest state
+		if err := sb.updateControlConfig(ctx, 0, 0); err != nil {
+			return err
+		}
```
kinda forgot that as a user I could just tell my controller to stop. so im gonna do that now. 

running into a small issue where if I have a sensorcontrolled base and controlled motor running, calling stop will break the motor, but I can probably track that down

martha-johnston replied:
is this issue resolved?

JohnN193 replied:
yep the issue is fixed now, the reason it was happening was because I set a velocity of 0. I didn't notice that at first because Next wasn't being locked, so reset would occur during next calls

JohnN193 commented on control/control_loop.go:
```diff
@@ -300,17 +300,19 @@ func (l *Loop) Stop() {
 
 // Resume sets l.running to true to resume the loop.
 func (l *Loop) Resume() {
+	if !l.running.Load() {
```
the other fix potential fix is to reset the PID blocks when a user goes to resume, rather than when they pause. After testing this worked fine, just needed to call resume after the setpoints got updated.

martha-johnston replied:
I tried resetting on resume which does work but I took it out because that means if you call set velocity and the base is moving, and then you call set velocity again with a new velocity, it will briefly pause. I think this is not the end of the world but I do see how it's not ideal

JohnN193 replied:
oh this only resets if the previous state was paused to avoid that behavior.

JohnN193 commented:
> calling stop will break the motor

turns out this happened because passing in a velocity of 0 to the trapezoid block causes the math to explode. It was difficult to see this behavior because Next did not lock the mutex while running, so the behavior would rarely appear while testing. I only realized there was no mutex when I managed to break the control loop from not having the signal exist while reset was being called. 

updated the code to not set a velocity of 0, and added some checks to ensure the velocity is never set to 0.

martha-johnston commented on components/base/sensorcontrolled/velocities.go:
```diff
@@ -30,12 +30,12 @@ func (sb *sensorBase) SetVelocity(
 	if err := sb.updateControlConfig(ctx, linear.Y/1000.0, angular.Z); err != nil {
 		return err
 	}
+	sb.loop.Resume()
```
does this cause a pause between set velocity calls in the same direction at different speeds? for example, set velocity at 100 mm/s and then set velocity at 200 mm/s, do the wheels stop before going to 200 or just speed up?

martha-johnston replied:
nvm I see the changes in resume

martha-johnston commented on control/trapezoid_velocity_profile.go:
```diff
@@ -107,8 +109,14 @@ func (s *trapezoidVelocityGenerator) reset() error {
+	s.maxAcc = s.cfg.Attribute["max_acc"].(float64)
+	s.maxVel = s.cfg.Attribute["max_vel"].(float64)
+	if s.maxAcc == 0 { // default 1.0, the math breaks if maxAcc = 0
+		s.maxAcc = 0
```
is this supposed to be `s.maxAcc = 1`?

martha-johnston commented on control/trapezoid_velocity_profile.go:
```diff
@@ -122,6 +130,10 @@ func (s *trapezoidVelocityGenerator) reset() error {
+	s.lastVelCmd = 0
+	if s.y != nil {
+		s.lastVelCmd = s.y[0].GetSignalValueAt(0)
+	}
```
I'm not clear on what this is doing

JohnN193 replied:
if the trapezoid block already has a desired velocity signal, we should preserve that state in `lastVelCmd`. this was required to persist velocity commands thru multiple `SetRPM` calls. It's possible the equations should be tweaked so that we no longer assume a non-zero velocity in the math, but I think this works for now.

martha-johnston reviewed (approved):
left a couple of questions but lgtm. also want to verify the hardware tests listed in the description all passed successfully
